easybuild/easyblocks/generic/pythonpackage.py

```diff
@@ -446,6 +446,8 @@ def build_step(self):
                 env.setvar("CMAKE_INCLUDE_PATH", include_paths)
                 env.setvar("CMAKE_LIBRARY_PATH", library_paths)
 
+            env.setvar("LDSHARED", "%s -shared" % self.toolchain.get_variable("CC", str))
```

Comment:
AFAIK this is necessary just when compiling python packages with `icc`, when python has been compiled in `GCCcore`. Possibly the same is true when compiling with PGI, but I've never tried it. I am basically worried about introducing side effects when applying this to every situation. I would suggest to include a check that verifies the compiler used in the toolchain of `Python` or `Python-core` and the compiler used in the toolchain being used to compile this particular package, and add `LDSHARED` just when both differ.

Reply:
Well, yes. By default, Pythons distutils will take the linker name from when the interpreter was compiled, regardless of what CC is set to in the package currently being built.

From what I can see in distutils source code, I couldn't spot any potential harm in always setting LDSHARED. If anything, it leads to less guesswork for distutils.
Here is a link to the relevant code;
https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/distutils/sysconfig.py#L172

I don't mind adding a conditional check though, but I don't do much hacking on blocks, so I would appreciate a hint on what's the cleanest way to check this would be.

Reply:
I sort of share @damianam's concern on always setting `$LDSHARED` here.
On the other hand, if both compilers are the same, setting `$LDSHARED` like this shouldn't have any impact. In theory...

I checedk what this spits out for a couple of different `Python` installations:

```
python -c "from distutils.sysconfig import get_config_vars; print(get_config_vars('LDSHARED'))[0]"
```

Results:

* system Python 2 (`/usr/bin/python`) in CentOS 7:
  ```
  gcc -pthread -shared -Wl,-z,relro
  ```
  This already disagrees with the hardcoded `$CC -shared` that is being added here.

* `Python/2.7.14-foss-2018a`
  ```
  gcc -shared -L/prefix/software/GCCcore/6.4.0/lib64 -L/prefix/software/GCCcore/6.4.0/lib -L/prefix/software/OpenBLAS/0.2.20-GCC-6.4.0-2.28/lib -L/prefix/software/ScaLAPACK/2.0.2-gompi-2018a-OpenBLAS-0.2.20/lib -L/prefix/software/FFTW/3.3.7-gompi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib
  ```
  This is more than just `$CC -shared`, it's also a bunch of `-L` specs, seems like one for each Python dependency (probably using `$LDFLAGS` that was set when `Python` was built).

* `Python/2.7.14-intel-2018a`
  ```
  icc -shared -L/prefix/software/icc/2018.1.163-GCC-6.4.0-2.28/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/imkl/2018.1.163-iimpi-2018a/mkl/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib
  ```
  Again `$CC -shared` + `-L` specs determined by `$LDFLAGS` that was set when building this `Python`.

* `Python/3.6.4-foss-2018a`
  ```
  gcc -shared -L/prefix/software/GCCcore/6.4.0/lib64 -L/prefix/software/GCCcore/6.4.0/lib -L/prefix/software/OpenBLAS/0.2.20-GCC-6.4.0-2.28/lib -L/prefix/software/ScaLAPACK/2.0.2-gompi-2018a-OpenBLAS-0.2.20/lib -L/prefix/software/FFTW/3.3.7-gompi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib -L/prefix/software/GCCcore/6.4.0/lib64 -L/prefix/software/GCCcore/6.4.0/lib -L/prefix/software/OpenBLAS/0.2.20-GCC-6.4.0-2.28/lib -L/prefix/software/ScaLAPACK/2.0.2-gompi-2018a-OpenBLAS-0.2.20/lib -L/prefix/software/FFTW/3.3.7-gompi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib -lssp
  ```
  Same as for `Python/2.7.14-foss-2018a`, except for the additional `-lssp` at the end (which is self-inflicted, we add `-lssp` to `$LDFLAGS` in easyconfigs for recent Python 3.x verisons).

* `Python/3.6.4-foss-2018a`
  ```
  icc -shared -L/prefix/software/icc/2018.1.163-GCC-6.4.0-2.28/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/imkl/2018.1.163-iimpi-2018a/mkl/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib -L/prefix/software/icc/2018.1.163-GCC-6.4.0-2.28/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/imkl/2018.1.163-iimpi-2018a/mkl/lib/intel64 -L/prefix/software/imkl/2018.1.163-iimpi-2018a/lib -L/prefix/software/bzip2/1.0.6-GCCcore-6.4.0/lib -L/prefix/software/zlib/1.2.11-GCCcore-6.4.0/lib -L/prefix/software/libreadline/7.0-GCCcore-6.4.0/lib -L/prefix/software/ncurses/6.0-GCCcore-6.4.0/lib -L/prefix/software/SQLite/3.21.0-GCCcore-6.4.0/lib -L/prefix/software/GMP/6.1.2-GCCcore-6.4.0/lib -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib64 -L/prefix/software/libffi/3.2.1-GCCcore-6.4.0/lib -lssp
  ```
  Basically the same as for `Python/3.6.4-foss-2018a`, but with `gcc` replaced by `icc` (also extra `-lssp` compared to `Python/2.7.14-intel-2018a`).

Long story short: always setting `$LDSHARED` to `$CC -shared` is probably not a good idea, since then we'd effectively be ignoring the `$LDFLAGS` part of it. But we can ask `Python` what it thinks `$LDSHARED` should be, and then replace the compiler command? (see below)

To make the setting of `$LDSHARED` conditional, we can check the value that `get_config_vars('LDSHARED')` spits out, and see whether it starts with the current value for `$CC`. Something like this (untested, but should work I think):

```python
# take into account that $LDSHARED may already be set...
curr_ldshared = os.getenv('LDSHARED')
python_ldshared = get_config_vars('LDSHARED')[0].split(' ')
curr_cc = os.getenv('CC')
if python_ldshared[0] != curr_cc:
    self.log.info("Python's value for $LDSHARED ('%s') doesn't use current $CC value ('%s'), fixing that...", python_ldshared, curr_cc)
    if curr_ldshared:
        # not sure what to do here... print/log a warning that $LDSHARED will be redefined? raise an error?
    env.setvar('LDSHARED', ' '.join([curr_cc] + python_ldshared[1:])
```

Not sure if the search-and-replace is wise. Maybe just setting it to `$CC -shared` conditionally is better/good enough?

Reply:
If so, my vote is to conditionally set it to "CC - shared".
Keeping the flags from another linker seems fragile, as there wouldn't really be any guarantee that the other compiler uses the same names (not the case here, but it was actually link flags compatibility differences which caused me to find this problem to start with when I started compiling for Python-core. (in that case, was just some `icc` compile flag that leaked into the link stage if i recall correctly, but it was enough to break the link step)

If there are no objections, I will update with
```python
python_ldshared = get_config_vars('LDSHARED')[0].split(' ')
curr_cc = os.getenv('CC')
if python_ldshared[0] != curr_cc:
    env.setvar("LDSHARED", curr_cc + " -shared")
```


Should i prefer using `os.getenv` or `toolchain.get_variable` here?

Reply:
@Micket I don't think it matters much which you use, `LDSHARED` will be set by the toolchain mechanism, so there's no point to try and grab it via `self.toolchain` imho...